From b1141091eb81b26009077cd6bb4920daa4139a2f Mon Sep 17 00:00:00 2001
From: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Date: Fri, 21 Nov 2025 18:30:26 +0100
Subject: [PATCH] usb: dwc2: drd: skip multiple role_sw notifications

The role_sw may be called several times, with the same data role.
This happens with ucsi controller, initial plug gives the role, later
notifications gets received upon src caps in device mode.

When the current role is already the expected one, skip it. Choice is
made to use a dedicated variable in hsotg struct, to properly manage
the "none" role (alternatively dwc2_is_[device|host]_mode() could have
been used, but isn't enough to manage the none role).

Add the check after the role_sw_default_mode override, to avoid
multiple entry/exit from power saving modes, for the same role.
This may happen when sequential "none" role (overridden with
"peripheral" for example) is followed by a "peripheral" notification.

Signed-off-by: Fabrice Gasnier <fabrice.gasnier@foss.st.com>
Change-Id: Icc6de79b700d685c7f9679f3c0dd9e6af02cde79
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/512980
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
---
 drivers/usb/dwc2/core.h | 3 +++
 drivers/usb/dwc2/drd.c  | 7 +++++++
 2 files changed, 10 insertions(+)

diff --git a/drivers/usb/dwc2/core.h b/drivers/usb/dwc2/core.h
index b0f1b9c52eda..c1f0c8f51217 100644
--- a/drivers/usb/dwc2/core.h
+++ b/drivers/usb/dwc2/core.h
@@ -15,6 +15,7 @@
 #include <linux/usb/gadget.h>
 #include <linux/usb/otg.h>
 #include <linux/usb/phy.h>
+#include <linux/usb/role.h>
 #include "hw.h"
 
 /*
@@ -865,6 +866,7 @@ struct dwc2_hregs_backup {
  *                      - USB_DR_MODE_HOST
  *                      - USB_DR_MODE_OTG
  * @role_sw:		usb_role_switch handle
+ * @current_role:	current usb_role, when using role_sw
  * @role_sw_default_mode: default operation mode of controller while usb role
  *			is USB_ROLE_NONE
  * @hcd_enabled:	Host mode sub-driver initialization indicator.
@@ -1065,6 +1067,7 @@ struct dwc2_hsotg {
 	enum usb_otg_state op_state;
 	enum usb_dr_mode dr_mode;
 	struct usb_role_switch *role_sw;
+	enum usb_role current_role;
 	enum usb_dr_mode role_sw_default_mode;
 	unsigned int hcd_enabled:1;
 	unsigned int gadget_enabled:1;
diff --git a/drivers/usb/dwc2/drd.c b/drivers/usb/dwc2/drd.c
index c7900a4dcf6a..0ad2d09810e7 100644
--- a/drivers/usb/dwc2/drd.c
+++ b/drivers/usb/dwc2/drd.c
@@ -185,6 +185,11 @@ static int dwc2_drd_role_sw_set(struct usb_role_switch *sw, enum usb_role role)
 			role = USB_ROLE_DEVICE;
 	}
 
+	if (hsotg->current_role == role) {
+		spin_unlock_irqrestore(&hsotg->lock, flags);
+		goto skip;
+	}
+
 	if ((IS_ENABLED(CONFIG_USB_DWC2_PERIPHERAL) ||
 	     IS_ENABLED(CONFIG_USB_DWC2_DUAL_ROLE)) &&
 	     dwc2_is_device_mode(hsotg) &&
@@ -215,12 +220,14 @@ static int dwc2_drd_role_sw_set(struct usb_role_switch *sw, enum usb_role role)
 		}
 	}
 
+	hsotg->current_role = role;
 	spin_unlock_irqrestore(&hsotg->lock, flags);
 
 	if (!already && hsotg->dr_mode == USB_DR_MODE_OTG)
 		/* This will raise a Connector ID Status Change Interrupt */
 		dwc2_force_mode(hsotg, role == USB_ROLE_HOST);
 
+skip:
 	if (!hsotg->ll_hw_enabled && hsotg->clk)
 		clk_disable_unprepare(hsotg->clk);
 
-- 
2.53.0

